JSC: Fix -Wreturn-type error/warning

csaavedra opened the pull request:
#### 9cbd8da8ae8dd844d53b6ee4000c99cc357e8ba1
<pre>
JSC: Fix -Wreturn-type error/warning
<a href="https://bugs.webkit.org/show_bug.cgi?id=247058">https://bugs.webkit.org/show_bug.cgi?id=247058</a>

Reviewed by Ross Kirsling.

* Source/JavaScriptCore/jit/Width.h:
(JSC::mask): Add missing return at the end of the method.

Canonical link: <a href="https://commits.webkit.org/256010@main">https://commits.webkit.org/256010@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3358c0dd42b13b62260b908bb04131ce21ed3bd1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94335 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3512 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/27263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104002 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164276 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3557 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31719 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86671 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/100004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100005 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2552 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80729 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29599 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84472 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/27263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72490 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/85552 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38114 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/27263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/80707 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35991 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/27263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27753 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4159 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39878 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/83373 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41827 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/27263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18829 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/JavaScriptCore/jit/Width.h:
```diff
@@ -93,6 +93,7 @@ inline constexpr uint64_t mask(Width width)
     case Width64:
         return 0xffffffffffffffffllu;
     case Width128:
+    default:
```
This doesn't make sense though—the `case`s are exhaustive so there can be no `default` path.

We should either match `bytesForWidth` above, or otherwise drop `case Width128:`.

csaavedra replied:
Compiler disagrees:

```
In file included from /app/webkit/Source/JavaScriptCore/b3/B3Type.h:32,
                 from /app/webkit/Source/JavaScriptCore/b3/B3Bank.h:30,
                 from /app/webkit/Source/JavaScriptCore/b3/B3Value.h:30,
                 from /app/webkit/Source/JavaScriptCore/b3/B3FixSSA.h:30,
                 from /app/webkit/Source/JavaScriptCore/b3/B3Generate.cpp:37,
                 from /app/webkit/WebKitBuild/Debug/JavaScriptCore/DerivedSources/unified-sources/UnifiedSource-23a5fd0e-5.cpp:1:
/app/webkit/Source/JavaScriptCore/jit/Width.h: In function ‘constexpr uint64_t JSC::mask(Width)’:
/app/webkit/Source/JavaScriptCore/jit/Width.h:99:1: error: control reaches end of non-void function [-Werror=return-type]
```

rkirsling replied:
No, I understand that the compiler is unhappy, but it's not because the `switch` isn't exhaustive; it's that the compiler is unwilling to count `case`s when checking whether the function as a whole has a consistent return type.

_Adding_ a `default` is incorrect for any human reader because there are only five possible paths, but we can _replace_ the last one with a `default`.

rkirsling reviewed (approved):
That works too. :smile:

csaavedra committed:
JSC: Fix -Wreturn-type error/warning
https://bugs.webkit.org/show_bug.cgi?id=247058

Reviewed by Ross Kirsling.

* Source/JavaScriptCore/jit/Width.h:
(JSC::mask): Add missing return at the end of the method.

Canonical link: https://commits.webkit.org/256010@main
